File: 백준/Gold/33802. Difference Maximization/Difference Maximization.py
import sys
def print(*args, sep=" ", end="\n"): return sys.stdout.write(sep.join(map(str, args)) + end)
def input(): return sys.stdin.readline().rstrip()

def main():
    n, m = map(int, input().split())
    A = [0]*(n+1); dmin = dmax = ans = 0
    for a in list(map(int, input().split())):
        if a:
            dmin += a-1
            dmax += m-a
        A[a] += 1
    s = 0
    for i in range(1, m+1):
        s += A[i]
        ans += s*(n-A[0]-s)
    while A[0]:
        if dmin < dmax:
            dmin += m-1
            ans += dmax
        else:
            dmax += m-1
            ans += dmin
        A[0] -= 1
    print(ans)
    return 0
if __name__ == '__main__':
    sys.exit(main())

File: 백준/Gold/33802. Difference Maximization/Difference Maximization.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int n , m; cin >> n >> m;
    vector<ll> A(m+1, 0); ll dmin = 0, dmax = 0;
    for(int i = 1; i <= n; i++){
        int a; cin >> a;
        if(a){
            dmin += a-1;
            dmax += m-a;
        }
        A[a]++;
    }
    ll s = 0, ans = 0;
    for(int i = 1; i < m; i++){
        s += A[i];
        ans += s*(n-A[0]-s);
    }
    while(A[0]--){
        if(dmin < dmax){
            dmin += m-1;
            ans += dmax;
        }
        else{
            dmax += m-1;
            ans += dmin;
        }
    }
    cout << ans;
    return 0;
}